### PR TITLE
show URL in viewer

### LINKED ISSUE #118: Display a URL in the console viewers
From the retrospective for https://github.com/Particular/Events/issues/73

The URL should be the best place to find out more about HashBus, including the location of the web UI.

Perhaps it should just point to the web UI, which can have link to the source, etc.

---

Will link to GH repo until web UI has more work done on it.

### PR DESCRIPTION
**Pull request by adamralph**
Fixes #118 

## Before

image

## After

image

### CODE REVIEW

File: src/HashBus.Viewer.MostMentioned/LeaderboardView.cs [hunk 1]
```diff
@@ -180,9 +182,9 @@ public Task RunAsync()
 
                     var tokens = new[]
                     {
-                        $"\r{padding}Total {itemsName}: ".DarkGray(),
-                        $"{currentLeaderboard?.Count ?? 0:N0}".Color(totalColor),
-                        $" · Refreshing in {timeLeft.Humanize()}...".DarkGray()
+                        $"\r{padding}Fork me at ".DarkGray(),
```
Comment (mikeminutillo) [inline src/HashBus.Viewer.MostMentioned/LeaderboardView.cs]: Is "Fork me at" necessary? If people are interested, is the URL alone enough?

File: src/HashBus.Viewer.MostMentioned/LeaderboardView.cs [hunk 2]
```diff
@@ -180,9 +182,9 @@ public Task RunAsync()
 
                     var tokens = new[]
                     {
-                        $"\r{padding}Total {itemsName}: ".DarkGray(),
-                        $"{currentLeaderboard?.Count ?? 0:N0}".Color(totalColor),
-                        $" · Refreshing in {timeLeft.Humanize()}...".DarkGray()
+                        $"\r{padding}Fork me at ".DarkGray(),
+                        "github.com/Particular/HashBus".Blue(),
+                        $" · {timeLeft.TotalSeconds}".DarkGray(),
```
Comment (mikeminutillo) [inline src/HashBus.Viewer.MostMentioned/LeaderboardView.cs]: If you dropped, fork me at does "Refresh" fit? Alternatively a symbol like ↻ might indicate the intent better.

File: src/HashBus.Viewer.MostMentioned/LeaderboardView.cs [hunk 3]
```diff
@@ -180,9 +182,9 @@ public Task RunAsync()
 
                     var tokens = new[]
                     {
-                        $"\r{padding}Total {itemsName}: ".DarkGray(),
-                        $"{currentLeaderboard?.Count ?? 0:N0}".Color(totalColor),
-                        $" · Refreshing in {timeLeft.Humanize()}...".DarkGray()
+                        $"\r{padding}Fork me at ".DarkGray(),
+                        "github.com/Particular/HashBus".Blue(),
```
Comment (mikeminutillo) [inline src/HashBus.Viewer.MostMentioned/LeaderboardView.cs]: I found the dark blue on black did not have enough contrast. It was difficult for me to read on both my screens.

File: src/HashBus.Viewer.MostMentioned/LeaderboardView.cs [hunk 4]
```diff
@@ -160,14 +159,17 @@ public Task RunAsync()
...
+
+                var totalColor = currentLeaderboard?.Count - previousLeaderboard?.Count > 0 ? movementColors[-1] : movementColors[0];
+                ColorConsole.Write(padding, $"{currentLeaderboard?.Count ?? 0:N0}".Color(totalColor));
+
                 if (currentLeaderboard.Since.HasValue)
                 {
                     ColorConsole.WriteLine(
-                        padding,
-                        $"{currentLeaderboard.Since?.ToLocalTime()} to {currentLeaderboard.LastActivityDateTime?.ToLocalTime()}".Gray());
+                        $" from {currentLeaderboard.Since?.ToLocalTime()} to {currentLeaderboard.LastActivityDateTime?.ToLocalTime()}".Gray());
```
Comment (mikeminutillo) [inline src/HashBus.Viewer.MostMentioned/LeaderboardView.cs]: There are really easy to gloss over. Given when we run these things, is it fair to say that:
- the year is unlikely to be relevant
- the month is probably also irrelevant
- the last activity date/time is likely to be nowish
- the seconds are likely to be irrelevant (especially given whatever latency is inherent to the system)

Would it make sense to have the message say something like:

> 642 from Tuesday 13:05 to now

**Comment by adamralph** [comment]
@mikeminutillo all very good points, thanks!

Regarding the date, even "to now" is redundant, so I gave "since xxxx" a try. The resulting shortening allowed me to reinstate the item type string, e.g. "tweets/retweets", "mentions", etc.

> Is "Fork me at" necessary? If people are interested, is the URL alone enough?

I was inspired by [the GitHub ribbons](https://github.com/blog/273-github-ribbons) but yes, it's redundant. I've removed it.

> I found the dark blue on black did not have enough contrast. It was difficult for me to read on both my screens.

It looked OK on my screen, but it's true that this won't work everywhere. In fact, now that I think of it, this is the reason that we made the user names, hashtags etc. (i.e. URL-like artifacts) cyan instead of blue. I've changed the URL to be cyan to fit that scheme.

> If you dropped, fork me at does "Refresh" fit? Alternatively a symbol like ↻ might indicate the intent better.

Absolutely. I've reinstated part of the text. As much as I'd like to use a Unicode symbol, we had problems with them in the past. We originally used Unicode arrows to indicate the movement of lines in the leaderboard, instead of `^`, `=`, and `v`, but for a reason I couldn't figure out, I could never get them to show properly on the Intel sticks, which would insist on showing a sentinel block character instead.

I've pushed a `fixup!` commit (to be squashed before merge) and this is how it looks now. What do you think?

image

File: src/HashBus.Viewer.MostMentioned/LeaderboardView.cs [hunk 5]
```diff
@@ -160,14 +159,18 @@ public Task RunAsync()
...
+                var totalColor = currentLeaderboard?.Count - previousLeaderboard?.Count > 0 ? movementColors[-1] : movementColors[0];
+                ColorConsole.Write(padding, $"{currentLeaderboard?.Count ?? 0:N0} ".Color(totalColor), $"{itemsName}".DarkGray());
+
                 if (currentLeaderboard.Since.HasValue)
                 {
                     ColorConsole.WriteLine(
-                        padding,
-                        $"{currentLeaderboard.Since?.ToLocalTime()} to {currentLeaderboard.LastActivityDateTime?.ToLocalTime()}".Gray());
+                        $" since ".DarkGray(),
+                        $"{currentLeaderboard.Since?.ToLocalTime():dddd} {currentLeaderboard.Since?.ToLocalTime():hh:mm}".Gray());
```
Comment (mikeminutillo) [inline src/HashBus.Viewer.MostMentioned/LeaderboardView.cs]: `HH` gives 24 hour time.
`hh` gives 12 hour time.

Given that we don't show AM/PM, I'd use 24 hour time

Comment (adamralph) [reply src/HashBus.Viewer.MostMentioned/LeaderboardView.cs]: Oops, good catch! Will fix.

**Comment by adamralph** [comment]
@mikeminutillo I've corrected the date format from `hh` to `HH` and squashed the commits. This should be good to go now.

**Comment by seanfarmar** [comment]
@adamralph @mikeminutillo ok to merge this?

**Comment by adamralph** [comment]
Merging, based on your approval @seanfarmar, and on the fact that @mikeminutillo's only outstanding request was to change from `hh` to `HH`, which has been done.

Thanks to you both!

**Pull request merged**